### PR TITLE
Batch 5 of Speedlimits UI: Changes to U library

### PR DESCRIPTION
**Pull request by kvakvs**
This is batch 5 instead of canceled huge batch 4.

This contains dropped in new files of the U library
* new atlases
* new button skins builder code
* simplified UI builder API
* new margins for UI elements, now all 4 sides of a panel can be configured separately
* all TM:PE code using it adapted to the new changes.

### CODE REVIEW

File: TLM/TLM/UI/MainMenu/MainMenuWindow.cs [hunk 1]
```diff
@@ -147,22 +147,22 @@ public class MainMenuWindow
         /// </summary>
```
Comment (krzychu124) [inline TLM/TLM/UI/MainMenu/MainMenuWindow.cs:147]: I can't add comment to untouched code but those fields are no longer used
```cs
private bool isStarted_;
//and
private MainMenuLayout menuLayout_;
```

File: TLM/TLM/UI/MainMenu/MainMenuWindow.cs [hunk 2]
```diff
@@ -537,13 +409,13 @@ private struct AddButtonsResult {
...
             }
+
             UResizer.UpdateControl(this);
         }
 
         public override void OnDestroy() {
             eventVisibilityChanged -= OnVisibilityChanged;
             confDisposable?.Dispose();
-            base.OnDestroy();
         }
```
Comment (krzychu124) [inline TLM/TLM/UI/MainMenu/MainMenuWindow.cs:419]: Why did you remove that method call? Without it `BaseUWindowPanel` will be leaking memory from not disposed observables.

File: TLM/TLM/UI/SubTools/LaneArrows/LaneArrowTool.cs [hunk 3]
```diff
@@ -179,23 +180,38 @@ private enum Trigger {
         /// </summary>
         /// <param name="numLanes">How many groups of buttons.</param>
         private void CreateLaneArrowsWindow(int numLanes) {
-            var parent = UIView.GetAView();
-            ToolWindow = (LaneArrowToolWindow)parent.AddUIComponent(typeof(LaneArrowToolWindow));
-            ToolWindow.SetOpacity(
-                U.UOpacityValue.FromOpacity(0.01f * GlobalConfig.Instance.Main.GuiOpacity));
+            // var parent = UIView.GetAView();
```
Comment (krzychu124) [inline TLM/TLM/UI/SubTools/LaneArrows/LaneArrowTool.cs:183]: Will that commented code be used in the future (next batch) or you forgot to remove it after refactoring?

File: TLM/TLM/UI/MainMenu/MainMenuWindow.cs [hunk 4]
```diff
@@ -15,10 +15,9 @@ namespace TrafficManager.UI.MainMenu {
     using UnityEngine;
     using TrafficManager.Lifecycle;
 
-    public class MainMenuWindow
+    public partial class MainMenuWindow
```
Comment (krzychu124) [inline TLM/TLM/UI/MainMenu/MainMenuWindow.cs:18]: I'm not a fan of partial classes since they are mostly for externally generated code (using some kind of UI designer or something)

File: TLM/TLM/UI/MainMenu/MainMenuWindow.ToolPanel.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                public MainMenuLayout Layout;
+            }
+
+            /// <summary>Create buttons and add them to the given panel UIBuilder.</summary>
+            /// <param name="builder">UI builder to use.</param>
+            /// <param name="atlasKeysSet">Atlas keys to update for button images.</param>
+            /// <param name="buttonDefs">Button defs collection to create from it.</param>
+            /// <param name="minRowLength">Longest the row can be without breaking.</param>
+            /// <returns>A list of created buttons.</returns>
+            private AddButtonsResult AddButtonsFromButtonDefinitions(
```
Comment (krzychu124) [inline TLM/TLM/UI/MainMenu/MainMenuWindow.ToolPanel.cs:22]: Missing params, `atlasKeysSet` does not exist

File: TLM/TLM/U/Panel/BaseUWindowPanel.cs [hunk 6]
```diff
@@ -45,13 +45,6 @@ public abstract class BaseUWindowPanel
...
         public virtual void OnAfterResizerUpdate() { }
 
-        /// <summary>Called by UnityEngine when component gets destroyed</summary>
-        public override void OnDestroy() {
-            uiScaleUnbsubscriber_.Dispose();
-            uiTransparencyUnbsubscriber_.Dispose();
-            base.OnDestroy();
-        }
-
         /// <summary>Invoke rescaling handler, because possibly it has the new size now.</summary>
```
Comment (krzychu124) [inline TLM/TLM/U/Panel/BaseUWindowPanel.cs:48]: Why removed?

**Review by krzychu124 — CHANGES_REQUESTED** [review]

**Comment by krzychu124** [comment]
Runtime problems:
### 1st: 
```cs
Debug 383.1107806: ##### Initializing ModUI.
Error 383.9304309: While creating MainMenu: System.IndexOutOfRangeException: Array index is out of range.
  at TrafficManager.State.ConfigData.DebugSwitchExtensions.Get (DebugSwitch sw) [0x00000] in <filename unknown>:0 
  at TrafficManager.U.Autosize.UResizerConfig.CallOnResize (ColossalFramework.UI.UIComponent control, ColossalFramework.UI.UIComponent previousSibling, UBoundingBox childrenBox) [0x00000] in <filename unknown>:0 
  at TrafficManager.U.Autosize.UResizer.UpdateControlRecursive (ColossalFramework.UI.UIComponent current, ColossalFramework.UI.UIComponent previousSibling) [0x00000] in <filename unknown>:0 
  at TrafficManager.U.Autosize.UResizer.UpdateControlRecursive (ColossalFramework.UI.UIComponent current, ColossalFramework.UI.UIComponent previousSibling) [0x00000] in <filename unknown>:0 
  at TrafficManager.U.Autosize.UResizer.UpdateControl (ColossalFramework.UI.UIComponent control) [0x00000] in <filename unknown>:0 
  at TrafficManager.U.Autosize.SmartResizeExtensions.ForceUpdateLayout (ColossalFramework.UI.UIComponent ctrl) [0x00000] in <filename unknown>:0 
  at TrafficManager.UI.MainMenu.MainMenuWindow.CreateMainMenuWindow () [0x00000] in <filename unknown>:0 
  at TrafficManager.UI.ModUI.CreateMainMenuButtonAndWindow () [0x00000] in <filename unknown>:0 
   at CSUtil.Commons.Log.LogToFile(System.String log, LogLevel level)
   at CSUtil.Commons.Log.Error(System.String s)
   at TrafficManager.UI.ModUI.CreateMainMenuButtonAndWindow()
   at TrafficManager.UI.ModUI.Awake()
   at UnityEngine.GameObject.Internal_AddComponentWithType(System.Type )
   at UnityEngine.GameObject.AddComponent(System.Type componentType)
   at UnityEngine.GameObject.AddComponent()
   at TrafficManager.UI.ModUI.OnLevelLoaded()
   at TrafficManager.Lifecycle.TMPELifecycle.Load()
   at TrafficManager.Lifecycle.TrafficManagerMod.OnLevelLoaded(LoadMode mode)
   at LoadingWrapper.OnLevelLoaded(UpdateMode mode)
   at LoadingManager+<LoadLevelComplete>c__Iterator9.MoveNext()
```
If you add new Debug flag to `GlobalConfig` you have to bump version to force config rebuild, otherwise it will throw exception as above.

### 2nd (TMPE.log spam):
```cs
Error 394.5187391: GUI Error: System.NullReferenceException: Object reference not set to an instance of an object
  at TrafficManager.UI.TrafficManagerTool.OnToolGUIImpl (UnityEngine.Event e) [0x00000] in <filename unknown>:0 
   at CSUtil.Commons.Log.LogToFile(System.String log, LogLevel level)
   at CSUtil.Commons.Log.Error(System.String s)
   at TrafficManager.UI.TrafficManagerTool.OnToolGUIImpl(UnityEngine.Event e)
   at TrafficManager.UI.TrafficManagerTool.OnToolGUI(UnityEngine.Event e)
   at ToolBase.OnGUI()
```
```cs
NullReferenceException: Object reference not set to an instance of an object
  at TrafficManager.UI.TrafficManagerTool.OnToolUpdate () [0x00000] in <filename unknown>:0 
```

File: TLM/TLM/U/Panel/BaseUWindowPanel.cs (hunk 6 above)
Comment (kvakvs) [reply TLM/TLM/U/Panel/BaseUWindowPanel.cs:48]: Asssumed garbage collector will eventually pick these up. Or do we have to do explicit destructions now?

File: TLM/TLM/UI/MainMenu/MainMenuWindow.cs (hunk 4 above)
Comment (kvakvs) [reply TLM/TLM/UI/MainMenu/MainMenuWindow.cs:18]: UI code grows long very quick, there isn't an UI designer file to hide the layout and setup code, so i am trying to not have it in the same file where the rest of the logic sits.

File: TLM/TLM/UI/MainMenu/MainMenuWindow.cs (hunk 1 above)
Comment (kvakvs) [reply TLM/TLM/UI/MainMenu/MainMenuWindow.cs:147]: Removed, and this also makes MainMenuLayout class unused, removed too

File: TLM/TLM/UI/MainMenu/MainMenuWindow.cs (hunk 2 above)
Comment (kvakvs) [reply TLM/TLM/UI/MainMenu/MainMenuWindow.cs:419]: Reverted this line

File: TLM/TLM/UI/SubTools/LaneArrows/LaneArrowTool.cs (hunk 3 above)
Comment (kvakvs) [reply TLM/TLM/UI/SubTools/LaneArrows/LaneArrowTool.cs:183]: This was old code before i copied in the new code. Was not sure whether it will just become a drop-in replacement, and it did. Removed the commented code.

File: TLM/TLM/UI/MainMenu/MainMenuWindow.ToolPanel.cs (hunk 5 above)
Comment (kvakvs) [reply TLM/TLM/UI/MainMenu/MainMenuWindow.ToolPanel.cs:22]: Updated XMLdoc

File: TLM/TLM/U/Panel/BaseUWindowPanel.cs (hunk 6 above)
Comment (kvakvs) [reply TLM/TLM/U/Panel/BaseUWindowPanel.cs:48]: Reverted

**Review by krzychu124 — APPROVED** [review]
Good 👍

File: TLM/CSUtil.Commons/Log.cs [hunk 7]
```diff
@@ -163,16 +163,9 @@ private enum LogLevel {
...
 
-        /// <summary>
-        /// Log an error only if cond is true
-        /// NOTE: If a lambda contains values from `out` and `ref` scope args,
-        /// then you can not use a lambda, instead use `if (cond) { Log.Error() }`
-        /// </summary>
-        /// <param name="cond">The condition</param>
-        /// <param name="s">The function which returns text to log</param>
-        public static void ErrorIf(bool cond, Func<string> s) {
+        public static void ErrorIf(bool cond, string s) {
```
Comment (kianzarrin) [inline TLM/CSUtil.Commons/Log.cs:166]: i think there the reason this accepted function instead of string was because the author was trying to avoid performing string formatting in debug builds.

are you sure this does not reduce performance or allocate garbage memory in if there is no error?

in any case the lambda to generate formatted string created garbage in the previous code anyway.

File: TLM/TLM/TrafficLight/Impl/TimedTrafficLights.cs [hunk 8]
```diff
@@ -1091,7 +1091,7 @@ public class TimedTrafficLights : ITimedTrafficLights {
             if (!Constants.ServiceFactory.NetService.IsSegmentValid(segmentId)) {
                 Log.ErrorIf(
                     logTrafficLights,
-                    () => $"TimedTrafficLights.ChangeLightMode: Segment {segmentId} is invalid");
+                    $"TimedTrafficLights.ChangeLightMode: Segment {segmentId} is invalid");
```
Comment (kianzarrin) [inline TLM/TLM/TrafficLight/Impl/TimedTrafficLights.cs:1094]: [low priority note]: 
String.format still creates garbage. I think the segmentId is passed to params [] and allocates array. 

I think IDE should give you a warning about this anyway.

In any case this does not look like performance critical part of the code to me so its fine.

File: TLM/TLM/U/Autosize/UPadding.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
+namespace TrafficManager.U.Autosize {
```
Comment (kianzarrin) [inline TLM/TLM/U/Autosize/UPadding.cs:1]: is there a reason you avoided using `UnityEngine.Rect`?

File: TLM/TLM/U/FreeTooltip/UFloatingTooltip.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
+﻿namespace TrafficManager.U {
+    using JetBrains.Annotations;
+    using UnityEngine;
+
+    /// <summary>
+    /// A free floating UI label which follows the mouse and displays current tool mode.
+    /// </summary>
+    public class UFloatingTooltip : ULabel {
```
Comment (kianzarrin) [inline TLM/TLM/U/FreeTooltip/UFloatingTooltip.cs:8]: Note: I have noticed that its possible to use `ToolBase.ShowToolInfo()` to show floating toolbox. 
 
This is ok too.

File: TLM/TLM/U/HelperExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        public static Rect GetScreenRectInGuiSpace(this UIComponent ctrl) {
+            Vector3 absPosition = ctrl.absolutePosition;
+            float scaleX = Screen.width / 1920f;
+            float scaleY = Screen.height / 1080f;
+            return new Rect(
+                absPosition.x * scaleX,
+                absPosition.y * scaleX,
+                ctrl.width * scaleY,
+                ctrl.height * scaleY);
+            // var b = ctrl.GetBounds();
```
Comment (kianzarrin) [inline TLM/TLM/U/HelperExtensions.cs:19]: did you forget to delete commented out code?

File: TLM/TLM/U/HelperExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+    public static class HelperExtensions {
+        /// <summary>
+        /// Get rectangle of a control in GUI screen space.
+        /// GUI screen space is used to draw overlay textures in overlay helpers.
+        /// </summary>
+        public static Rect GetScreenRectInGuiSpace(this UIComponent ctrl) {
+            Vector3 absPosition = ctrl.absolutePosition;
+            float scaleX = Screen.width / 1920f;
+            float scaleY = Screen.height / 1080f;
```
Comment (kianzarrin) [inline TLM/TLM/U/HelperExtensions.cs:13]: does this work if screen aspect ratio changes?

File: TLM/TLM/U/HelperExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
+﻿namespace TrafficManager.U {
+    using ColossalFramework.UI;
+    using UnityEngine;
+
+    public static class HelperExtensions {
+        /// <summary>
+        /// Get rectangle of a control in GUI screen space.
+        /// GUI screen space is used to draw overlay textures in overlay helpers.
+        /// </summary>
+        public static Rect GetScreenRectInGuiSpace(this UIComponent ctrl) {
```
Comment (kianzarrin) [inline TLM/TLM/U/HelperExtensions.cs:10]: I see no calls to this method and I have no way of testing this.
is this redundant or is it basis for future code?

**Comment by kianzarrin** [comment]
can please you explain how to test this code?

File: TLM/TLM/U/TextureUtil.cs [hunk 14]
```diff
@@ -22,19 +24,32 @@ public static class TextureUtil {
                                                  IntVector2 atlasSizeHint) {
             var loadedTextures = new List<Texture2D>(spriteDefs.Length);
             var loadedSpriteNames = new List<string>();
+#if DEBUG
+            bool debugResourceLoading = DebugSwitch.ResourceLoading.Get();
```
Comment (kianzarrin) [inline TLM/TLM/U/TextureUtil.cs:28]: should this be a property? what if user updates debug swiches mid game?

File: TLM/TLM/U/UBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    using ColossalFramework.UI;
+    using JetBrains.Annotations;
+    using TrafficManager.State.Keybinds;
+    using TrafficManager.U.Autosize;
+    using TrafficManager.UI;
+    using TrafficManager.Util;
+    using UnityEngine;
+
+    /// <summary>
+    /// Create an UI builder to populate a panel with good things: buttons, sub-panels, create a
```
Comment (kianzarrin) [inline TLM/TLM/U/UBuilder.cs:13]: good things lol!

File: TLM/TLM/U/Autosize/UPadding.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+                Bottom = 0f,
+            };
+        }
+
+        /// <summary>Set padding on all sides to <see cref="UConst.UIPADDING"/> (4f).</summary>
+        public static UPadding Const() {
+            return SameValue(UConst.UIPADDING);
+        }
+
+        public static UPadding SameValue(float v) {
```
Comment (kianzarrin) [inline TLM/TLM/U/Autosize/UPadding.cs:22]: why is this method necessary? strucs are copied by value anyway.
if this was a class I can understand the need for a clone function but why do we need this for such a simple struct?

File: TLM/TLM/U/Autosize/UPadding.cs [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+            return new() {
+                Right = 0f,
+                Top = 0f,
+                Left = 0f,
+                Bottom = 0f,
+            };
+        }
+
+        /// <summary>Set padding on all sides to <see cref="UConst.UIPADDING"/> (4f).</summary>
+        public static UPadding Const() {
```
Comment (kianzarrin) [inline TLM/TLM/U/Autosize/UPadding.cs:18]: is it possible to have a more meaningful name than Const?

File: TLM/TLM/U/UBuilder.cs [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+            window.SetPadding(UPadding.Const());
+
+            return window;
+        }
+
+        /// <summary>
+        /// Create a button as a child of current UIBuilder. A new UIBuilder is returned.
+        /// </summary>
+        /// <typeparam name="TButton">The type of UIButton and ISmartSizableControl.</typeparam>
+        public TButton Button<TButton, TParent>(TParent parent)
```
Comment (kianzarrin) [inline TLM/TLM/U/UBuilder.cs:56]: TParent is redundant. just use UIComponent
unless if you plan to add more restriction in the future.
does the parent also needs to be ISmartSizableControl?

Comment (kianzarrin) [reply TLM/TLM/U/UBuilder.cs:56]: please remove all redundant Generic arguments in the rest of the file too.

File: TLM/TLM/U/UConst.cs [hunk 19]
```diff
@@ -1,17 +1,28 @@
...
-    using UnityEngine;
-
     public static class UConst {
-        // private static readonly Color SHORTCUT_DESCR_TEXT = new Color(.75f, .75f, .75f, 1f);
-        // public static readonly Color SHORTCUT_KEYBIND_TEXT = new Color(.8f, .6f, .3f, 1f);
-
         /// <summary>Orange color for shortcut text.</summary>
         public static readonly string SHORTCUT_TEXT_HEX = "#cc994c";
 
+        public static string KEYBOARD_SHORTCUT_CLOSING_TAG = "</color>";
```
Comment (kianzarrin) [inline TLM/TLM/U/UConst.cs:6]: why these stings are not const?

File: TLM/TLM/U/UIScaler.cs [hunk 20]
```diff
@@ -14,7 +13,7 @@ public static class UIScaler {
         /// <summary>Caching because UIView.Instance.uiCamera can be null sometimes.</summary>
         private static float cachedGuiHeight = GUI_HEIGHT;
 
-        /// <summary>Screen width for GUI is always fixed at 1920.</summary>
+        /// <summary>Gets screen width for GUI (always fixed at 1920).</summary>
```
Comment (kianzarrin) [inline TLM/TLM/U/UIScaler.cs:16]: what if there is different aspect ratio? 
see https://github.com/kianzarrin/Skylines-ModTools/blob/master/Debugger/UI/UIScaler.cs
There is some guard code because the resolution is not ready at startup.

File: TLM/TLM/U/UIScaler.cs [hunk 21]
```diff
@@ -14,7 +13,7 @@ public static class UIScaler {
...
         public static float GuiWidth {
             // TODO: Double check if GUI never changes width, the code below can be a const
```
Comment (kianzarrin) [inline TLM/TLM/U/UIScaler.cs:18]: This is not called. is this redundant?

File: TLM/TLM/U/HelperExtensions.cs [hunk 22]
```diff
@@ -0,0 +1,27 @@
+﻿namespace TrafficManager.U {
+    using ColossalFramework.UI;
+    using UnityEngine;
+
+    public static class HelperExtensions {
+        /// <summary>
+        /// Get rectangle of a control in GUI screen space.
+        /// GUI screen space is used to draw overlay textures in overlay helpers.
+        /// </summary>
```
Comment (kianzarrin) [inline TLM/TLM/U/HelperExtensions.cs:9]: might be wise to use UIScaler to avoid duplicate code.

File: TLM/TLM/UI/MainMenu/ClearTrafficButton.cs [hunk 23]
```diff
@@ -14,13 +14,11 @@ public class ClearTrafficButton : BaseMenuButton {
...
         public override void SetupButtonSkin(AtlasBuilder futureAtlas) {
             // Button backround (from BackgroundPrefix) is provided by MainMenuPanel.Start
-            this.Skin = new U.ButtonSkin() {
-                Prefix = "ClearTraffic",
-                BackgroundPrefix = "RoundButton",
-                BackgroundHovered = true,
-                BackgroundActive = true,
-                ForegroundActive = true,
-            };
+            this.Skin = ButtonSkin.CreateSimple(
```
Comment (kianzarrin) [inline TLM/TLM/UI/MainMenu/ClearTrafficButton.cs:17]: indentation

File: TLM/CSUtil.Commons/Log.cs (hunk 7 above)
Comment (kvakvs) [reply TLM/CSUtil.Commons/Log.cs:166]: Both versions with lambda and with `$""` create garbage and lambda will do less work in case when condition is false.
I can probably revert it to Lambda variant.
This function is not used in critical code paths, only used in 2 locations (Timed TL: ChangeLightMode, and in U UI library if an exception happened during a resize event.

File: TLM/TLM/U/Autosize/UPadding.cs (hunk 9 above)
Comment (kvakvs) [reply TLM/TLM/U/Autosize/UPadding.cs:1]: Different semantic. Rect has x, y, width, height. And padding has left, right, top, bottom.

File: TLM/TLM/U/FreeTooltip/UFloatingTooltip.cs (hunk 10 above)
Comment (kvakvs) [reply TLM/TLM/U/FreeTooltip/UFloatingTooltip.cs:8]: I would probably prefer dropping this class and using the existing.
That API requires world position, this API works in screen coords.

File: TLM/TLM/U/HelperExtensions.cs (hunk 11 above)
Comment (kvakvs) [reply TLM/TLM/U/HelperExtensions.cs:19]: Deleted now. This class is currently shown in grey as unused.
It will be used after future batches are merged.

File: TLM/TLM/U/HelperExtensions.cs (hunk 12 above)
Comment (kvakvs) [reply TLM/TLM/U/HelperExtensions.cs:13]: GUI canvas seems to always maintain1920x1080 screen ratio internally. If this is not true, i need more info.

File: TLM/TLM/U/HelperExtensions.cs (hunk 13 above)
Comment (kvakvs) [reply TLM/TLM/U/HelperExtensions.cs:10]: Is used in future code, that is not merged in this PR. This PR merges entire new U directory so some pieces are here but not used yet.

File: TLM/TLM/U/TextureUtil.cs (hunk 14 above)
Comment (kvakvs) [reply TLM/TLM/U/TextureUtil.cs:28]: I have new version for this module to be merged with named properties.
It will be merged in a later PR
https://github.com/kvakvs/TMPE/blob/u-speedlimits-tool-4/TLM/TLM/State/ConfigData/DebugSettings.cs
If the user wants to debug texture loading they better set this switch before the loading?

File: TLM/TLM/U/UBuilder.cs (hunk 15 above)
Comment (kvakvs) [reply TLM/TLM/U/UBuilder.cs:13]: Bad things not allowed 😁

File: TLM/TLM/U/Autosize/UPadding.cs (hunk 16 above)
Comment (kvakvs) [reply TLM/TLM/U/Autosize/UPadding.cs:22]: This is not a method, is a constructor.
This is only used once but i believe the future code will use it a bit more.
Probably this can be rewritten with a const struct, will that be any better?

File: TLM/TLM/U/Autosize/UPadding.cs (hunk 17 above)
Comment (kvakvs) [reply TLM/TLM/U/Autosize/UPadding.cs:18]: Changed to 
```csharp
        public static UPadding Default = new UPadding() {
            Top = UConst.UIPADDING, Right = UConst.UIPADDING, Bottom = UConst.UIPADDING,
            Left = UConst.UIPADDING,
        };

        public static UPadding Zero = new UPadding() {
            Top = 0f, Right = 0f, Bottom = 0f, Left = 0f,
        };
```

File: TLM/TLM/U/UBuilder.cs (hunk 18 above)
Comment (kvakvs) [reply TLM/TLM/U/UBuilder.cs:56]: Done. Nice catch, thank you

File: TLM/TLM/U/UConst.cs (hunk 19 above)
Comment (kvakvs) [reply TLM/TLM/U/UConst.cs:6]: Now they are. Thanks

File: TLM/TLM/U/UIScaler.cs (hunk 20 above)
Comment (kvakvs) [reply TLM/TLM/U/UIScaler.cs:16]: This code exists for several releases, i remember i've done some testing and found that UI canvas size and aspect stays 1920x1080 at all times. If that's not true, i will need more info.

File: TLM/TLM/U/UIScaler.cs (hunk 21 above)
Comment (kvakvs) [reply TLM/TLM/U/UIScaler.cs:18]: GuiWidth property is used from UResizer.Calculate
cachedGuiHeight is used below in GuiHeight property

File: TLM/TLM/UI/MainMenu/ClearTrafficButton.cs (hunk 23 above)
Comment (kvakvs) [reply TLM/TLM/UI/MainMenu/ClearTrafficButton.cs:17]: Looks correct to me? 
bild

File: TLM/TLM/U/HelperExtensions.cs (hunk 22 above)
Comment (kvakvs) [reply TLM/TLM/U/HelperExtensions.cs:9]: Changed to 
```csharp
        public static Rect GetScreenRectInGuiSpace(this UIComponent ctrl) {
            Vector2 pos = UIScaler.ScreenPointToGuiPoint(ctrl.absolutePosition);
            Vector2 size = UIScaler.ScreenPointToGuiPoint(ctrl.size);
            return new Rect(pos, size);
        }
```

File: TLM/TLM/U/Autosize/UPadding.cs (hunk 16 above)
Comment (kvakvs) [reply TLM/TLM/U/Autosize/UPadding.cs:22]: Replaced with 2 constant structs of 0 and UConst.UIPADDING.

**Comment by kvakvs** [comment]
> can please you explain how to test this code?

This code does not add new features, but changes a lot of UI code.
See the window panels which use U library to be built: They are main game menu and lane arrows window, also small buttons are constructed with U library in street properties window when you click a street.

File: TLM/TLM/TrafficLight/Impl/TimedTrafficLights.cs (hunk 8 above)
Comment (kvakvs) [reply TLM/TLM/TrafficLight/Impl/TimedTrafficLights.cs:1094]: Reverted to use format lambda, it was better this way

File: TLM/CSUtil.Commons/Log.cs (hunk 7 above)
Comment (kvakvs) [reply TLM/CSUtil.Commons/Log.cs]: Reverted to lambda

File: TLM/TLM/U/UBuilder.cs [hunk 24]
```diff
@@ -17,13 +17,14 @@ namespace TrafficManager.U {
     public class UBuilder {
         public AtlasBuilder AtlasBuilder;
```
Comment (DaEgi01) [inline TLM/TLM/U/UBuilder.cs:18]: why is this a field and public?
is there a place you want to override it after instantiation?
if you can set it via ctor, it should be sufficient I assume.

File: TLM/TLM/U/UConst.cs [hunk 25]
```diff
@@ -1,9 +1,9 @@
 namespace TrafficManager.U {
     public static class UConst {
         /// <summary>Orange color for shortcut text.</summary>
-        public static readonly string SHORTCUT_TEXT_HEX = "#cc994c";
+        private const string SHORTCUT_TEXT_HEX = "#cc994c";
```
Comment (DaEgi01) [inline TLM/TLM/U/UConst.cs:4]: I'd have done those values static readonly because its not like these are values that can never change.
But for all intends and purposes it should not matter for us. if it breaks another mod in the future that depends on TM:PE, no big deal ^^

**Review by DaEgi01 — APPROVED** [review]

**Review by chameleon-tbn — APPROVED** [review]
Tested ingame, all looks good for me, no errors in logs

**Review by kianzarrin — APPROVED** [review]
lgtm
